### DISCUSSION TITLE
RDF4J SPIN Inference help

### DISCUSSION THREAD

**Discussion by bethhuffer** [eclipse-rdf4j/rdf4j#3208] [Q&A] [answered]
We have configured RDF4J as indicated below, but can't get inferences to work. Our ontology model includes SPIN rules that aren't firing, and even putting them directly into the Update window doesn't do the trick. For example, I would expect this:

```sparql
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX AMP: <http://lingualogica.net/AMP#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX spin: <http://spinrdf.org/spin#>
PREFIX sp: <http://spinrdf.org/sp#>
WITH <http://lingualogica.net/AMP>
INSERT {
	?x AMP:instrument AMP:CERESInstrument }
WHERE {
 	?x rdf:type AMP:CER_SSF1deg-Day_Aqua-MODIS_Edition4A 
}
```

to insert an assertion for the single test instance of AMP:CER_SSF1deg-Day_Aqua-MODIS_Edition4A, but it doesn't. We can't figure out where we've gone wrong. Any ideas? Thanks!

```turtle
@prefix ms: <http://www.openrdf.org/config/sail/memory#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix rep: <http://www.openrdf.org/config/repository#> .
@prefix sail: <http://www.openrdf.org/config/sail#> .
@prefix sb: <http://www.openrdf.org/config/sail/base#> .
@prefix sr: <http://www.openrdf.org/config/repository/sail#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<#memory-spin-rdfs> a rep:Repository;
rep:repositoryID "memory-spin-rdfs";
rep:repositoryImpl [
rep:repositoryType "openrdf:SailRepository";
sr:sailImpl [
sail:delegate [
sail:delegate [
sail:delegate [
sail:sailType "openrdf:MemoryStore";
sb:evaluationStrategyFactory "org.eclipse.rdf4j.query.algebra.evaluation.impl.StrictEvaluationStrategyFactory";
ms:persist true
];
sail:sailType "openrdf:DedupingInferencer"
];
sail:sailType "rdf4j:SchemaCachingRDFSInferencer"
];
sail:sailType "openrdf:SpinSail"
]
];
rdfs:label "Memory store with RDFS and full SPIN support" .
```

**Comment by abrokenjester** [comment]
Can you share what your actual SPIN rule (in the context of your ontology) looks like? You've shared a SPARQL INSERT operation, but SPIN rules are typically expressed as CONSTRUCT queries I believe.

Secondly, can you talk us through how you test things? With the SPIN Sail, I believe it is important to have the rule in place in the database _before_ you add the data that is supposed to trigger the rule.

Finally, I hope you are aware that SPIN support is deprecated, and is scheduled to be removed in RDF4J 4.0. See the [big orange warning in the documentation](https://rdf4j.org/documentation/programming/spin/).

**Reply by bethhuffer** [reply]
Thank you for your response. Yes, we have several rules like the one below, that are attributes of classes. 

AMP:CER_SSF1deg-Day_Aqua-MODIS_Edition4A
  rdf:type owl:Class ;
spin:constructor [
      rdf:type sp:Construct ;
      sp:templates (
          [
            sp:object AMP:CERESInstrument ;
            sp:predicate AMP:instrument ;
            sp:subject spin:_this  ;
              ] ;
          ]
        ) ;
      sp:where () ;
    ] ;

The ontology was developed in TopBraid Composer, and tested there by creating the classes, asserting the rules, and adding instances of the class. Then, we run a SPARQL construct query as follows:

CONSTRUCT { ?x ?y ?z }
WHERE {
   ?x rdf:type AMP:CER_SSF1deg-Day_Aqua-MODIS_Edition4A .
   ?y rdf:type AMP:AMPProperty .
   ?x ?y ?z 
}

Using TopBraid Composer, this generates the desired triples. But I can't get it to work in RDF4J. (I am aware, and also sad, that SPIN is deprecated. I don't mind converting these to SHACL rules, assuming I can make it work in the same way that SPIN works, but I haven't yet had time to figure out how to do that, and I'd like to be able to get this particular application working before I set about trying to do the conversion. Deadlines loom. Will RDF4J 4.0 provide support for user-defined rules?) 

Thanks again for your help!

**Reply by bethhuffer** [reply]
Responding to your question about when the rule is added: the rule gets inserted at the same time the class is created, using an INSERT DATA update statement. The instances are created in the same way - using INSERT DATA statements - after the class and the rule are inserted. So the rule is in place when the data is inserted, but it's not firing, either at insert time, or upon running the CONSTRUCT query I posted above.

**Comment by hmottestad** [comment, marked as answer]
Looks like you are adding data to an existing resource in your database. This is not supported by the SpinSail. 

Do you see the same results if you start with an empty repository, add your spin rules in one transaction then add all your data in another transaction?

What use case do you have for SPIN? I am always interested to hear about people who are still using it!

**Reply by bethhuffer** [reply]
We're trying to auto-generate metadata records for Earth science datasets. AMP=Automated Metadata Pipeline. The ontology/triple store is part of a larger data processing pipeline that parses a sample data file and generates a class representing a collection of datasets that share certain common attributes (e.g., the instrument that produced the data). Individual datasets are created as instances of that class, and SPIN rules are used to generate a set of assertions about each dataset that can be used to drive a data discovery and access service, that can also provide machine-readable, semantically consistent, robust metadata records to climate models, and other downstream analysis tools. Hoping that in this way we can make the data usable in automated, machine-to-machine transactions. The pipeline is working as we hoped, except that I need to get the ontology/triple store part off my desktop and into AWS in a way that preserves the functionality that I get using TopBraid Composer.

**Reply by hmottestad** [reply]
Perhaps TopQuadrant has a product you can purchase? SPIN is their specification after all.

**Reply by bethhuffer** [reply]
Yes, so it seems. Thanks for weighing in!
